build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java

```diff
@@ -199,5 +201,29 @@ public void execute(Task t) {
                 }
             });
         });
+        configureImmutableCollectionsPatch(project);
+    }
+
+    private void configureImmutableCollectionsPatch(Project project) {
+        String patchProject = ":test:immutable-collections-patch";
+        if (project.findProject(patchProject) == null) {
+            return; // build tests may not have this project, just skip
+        }
+        String configurationName = "immutableCollectionsPatch";
+        FileCollection patchedFileCollection = project.getConfigurations()
+            .create(configurationName, config -> config.setCanBeConsumed(false));
+        var deps = project.getDependencies();
+        deps.add(configurationName, deps.project(Map.of("path", patchProject, "configuration", "patch")));
+        project.getTasks().withType(Test.class).matching(task -> task.getName().equals("test")).configureEach(test -> {
+            test.getInputs().files(patchedFileCollection);
+            test.systemProperty("tests.hackImmutableCollections", "true");
+            test.getJvmArgumentProviders()
```

Comment:
FWIW, now that we are restricting this to the `test` task, using an argument provider doesn't really make sense. My original suggestion to do this during our conversation is that the argument provider could return a `List` conditional on the configured toolchain of the task, but this is just return a static list, so I don't see much point in doing this vs just configuring the JVM args directly.

Reply:
How would we delay resolving the configuration without using the argument provider?

Reply:
Ah, you are right. I forgot that's the other reason we want it to be lazy. 👍